lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -1419,28 +1464,78 @@ func (d *zfs) SetVolumeQuota(vol Volume, size string, allowUnsafeResize bool, op
 			return nil
 		}
 
-		// Block image volumes cannot be resized because they have a readonly snapshot that doesn't get
```

Comment:
@monstermunchkin I'm worried that this check isn't being enforced for filesystem block volumes (and no comment explaining that its been considered).

I'm observing that when creating a container using the default 10GiB size first, the optimized image block volume is created at 10GiB, and then a readonly snapshot is taken of that, as expected.

But when subsequent containers are created from the same image with root disks that are smaller than the optimized image volume, we are still using the optimized image volume snapshot as the origin for the new instance volume and then trying to shrink the volume and the filesystem to the new smaller size.

With the way this PR is currently, it appears to work, but I worry it'll cause issues if we try to create a volume that is smaller from an image volume snapshot that is larger (and actually is using the space). 

I think what is happening is that a 10GiB volume is safely created from the snapshot, then the filesystem is shrunk, and then the new volume is shrunk as well. But I need to do more testing to confirm this (I've run out of time today), and understand what ZFS is doing under the hood in these cases.

I also need to test what happens when using the XFS filesystem (which cannot be shrunk).

Reply:
> I also need to test what happens when using the XFS filesystem (which cannot be shrunk).

I can confirm this works.

```bash
lxc storage show default
config:
  size: 30GiB
  source: /var/lib/lxd/disks/default.img
  volume.block.filesystem: xfs
  volume.zfs.block_mode: "true"
  zfs.pool_name: default
  ...

lxc init images:alpine/3.17 c1

zfs get volsize default/containers/c1
NAME                   PROPERTY  VALUE    SOURCE
default/containers/c1  volsize   10G      local

lxc init images:alpine/3.17 c2 -d root,size=5GiB

zfs get volsize default/containers/c2
NAME                   PROPERTY  VALUE    SOURCE
default/containers/c2  volsize   5G       local
```

The log shows

```
Cached image volume is larger than new volume and cannot be shrunk, creating non-optimized volume
```

Reply:
Great thanks!